promql/functions.go

```diff
@@ -79,22 +75,29 @@ func funcDelta(ev *evaluator, args Expressions) model.Value {
 		}
 		resultValue := lastValue - samples.Values[0].Value + counterCorrection
 
-		targetInterval := args[0].(*MatrixSelector).Range
-		sampledInterval := samples.Values[len(samples.Values)-1].Timestamp.Sub(samples.Values[0].Timestamp)
-		if sampledInterval == 0 {
-			// Only found one sample. Cannot compute a rate from this.
-			continue
+		// Duration between first/last samples and boundary of range.
+		durationToStart := samples.Values[0].Timestamp.Sub(rangeStart).Seconds()
+		durationToEnd := rangeEnd.Sub(samples.Values[len(samples.Values)-1].Timestamp).Seconds()
+
+		sampledInterval := samples.Values[len(samples.Values)-1].Timestamp.Sub(samples.Values[0].Timestamp).Seconds()
+		averageDurationBetweenSamples := sampledInterval / float64(len(samples.Values))
+
+		// If the first/last samples are close to the boundaries of the range,
+		// extrapolate the result. This is as we expect that another sample
+		// will exist given the spacing between samples we've seen thus far,
+		// with an allowance for noise.
+		extrapolationThreshold := averageDurationBetweenSamples * 1.1
+		extrpolateToInterval := sampledInterval
+		if durationToStart < extrapolationThreshold {
+			extrpolateToInterval += durationToStart
+		}
+		if durationToEnd < extrapolationThreshold {
+			extrpolateToInterval += durationToEnd
+		}
```

Comment:
> Assuming a series starts at its first sample and ends at its last is way more bold than what I suggested.

That's only one set of assumptions though, there other ways that a series could be used as we need the rate function to behave reasonably for all of them as it's a generic function. We've already had a few users through for whom the last sample is the end of the series for example.

> Here we are applying heuristics, so we should stay consistent and not switch to something that is fundamentally different (as discussed at length before).

This is making it more and more difficult to figure out what this function is going to do. We should only do this sort of magic where we're sure it'll get the right answer in almost all cases, so that the average user doesn't have to end up reverse-engineering all this code which seems to be getting ever more complicated.

The purpose of this PR is to reduce the severity of the existing artifacts, let's let this settle in for a bit with our users before adding the next layer of epicycles.